### PR TITLE
Handle overlapping `members` and `exclude` keys

### LINKED ISSUE #4089: workspaces: `exclude` not working in Rustbuild
See https://github.com/rust-lang/rust/pull/42146#discussion_r117753454

### PR DESCRIPTION
**Pull request by alexcrichton**
The longest key takes precedence if there's more than one that matches.

Closes #4089

### CODE REVIEW

**Comment by alexcrichton** [comment]
Excellent point about globs! I fixed using `exclude` and globs by using `Gitignore`, although the `glob` crate is still used to expand `members` directives, so I [added a comment](https://github.com/rust-lang/cargo/issues/4268#issuecomment-316422019) on the issue to switch to gitignore.

**Comment by matklad** [comment]
Let's add a test with globs! :)

I think the scenario which does not work with old is_excluded is this:

Workspace root includes a member foo via glob. Then, if we execute cargo build *inside* foo, cargo won't detect that it is a member of the workspace.

File: src/cargo/core/workspace.rs [hunk 1]
```diff
@@ -578,25 +581,39 @@ fn expand_member_path(path: &Path) -> CargoResult<Vec<PathBuf>> {
...
-    });
+    let exclude_match = exclude.matched_path_or_any_parents(manifest_path, false);
 
     let explicit_member = match *members {
         Some(ref members) => {
-            members.iter().any(|mem| {
-                manifest_path.starts_with(root_path.join(mem))
-            })
+            members.iter()
+                .filter(|p| manifest_path.starts_with(root_path.join(p)))
```
Comment (matklad) [inline src/cargo/core/workspace.rs]: I think that members can contain globs. That is, `p` might be like `foo/*` here, and so this `starts_with` will fail.

**Comment by matklad** [comment]
>I think the scenario which does not work with old is_excluded is this:

I though wrong :) Nevertheless, I think this test will be interesting:

```
~/trash/ws master*
λ tree .
.
├── Cargo.toml
└── foo
    └── baz
        └── bar
            ├── Cargo.toml
            └── src
                └── lib.rs

4 directories, 3 files

~/trash/ws master*
λ cat Cargo.toml
[workspace]
members = ["foo/baz/*"]
exclude = ["foo/"]
```

Looks like `foo/bar/baz` should be a member of the workspace.

**Comment by behnam** [comment]
@alexcrichton, should I take over this one and fix https://github.com/rust-lang/cargo/issues/4089 along the migration to gitignore-matching everywhere?

**Comment by behnam** [comment]
Also, another difference between `workspace.member/exclude` vs `package.include/exclude` is that the `package` configs are exclusive (if `include` is set, `exclude` is ignored silently), but in `workspace`, turns out both configs can exist at the same time.

I think whatever solution we take here better be aligned with the future of `package` configs, since I'm hopping we can make those non-exclusive, too.

Here's my suggestion for `workspace` configs (which we can implement earlier and later expand it to `package` configs):

* Treat `member` and `exclude` configs as rows of one `gitignore`, with list in `member` being added with a `!` prefix.

* Error on negated patterns (patterns starting with `!`) in any of the configs. (Only allow escaped, literal exclamation mark: `\!`.)

What do you think?

**Comment by alexcrichton** [comment]
Sorry yeah @behnam if you wouldn't mind taking this over, that'd be great! I'm quickly losing steam on this :(

I'm also sorry that there's not really a great design here, it's all sort of ad-hoc. My main motivation for this was to remove [these lines in the rust-lang/rust repo](https://github.com/rust-lang/rust/blob/5c126bdee6af3feef2d51956debab72f933078c6/src/Cargo.toml#L20-L33). Beyond that I've just been attempting to rationalize gitignore/glob pattern, but I get the feeling that it's just becoming more of a mess.

**Comment by behnam** [comment]
No worries, @alexcrichton. I can get back to this when I get back from the trip next week.

**Comment by alexcrichton** [comment]
ping @behnam, was just curious if you had a chance to take a look at this again! If you're busy then of course no worries!

**Comment by alexcrichton** [comment]
This never quite panned out, so closing.

**Pull request closed without merging**